probe_src/probe_frontend/src/ops.rs

```diff
@@ -0,0 +1,410 @@
+#[allow(unused_imports)]
+pub use crate::ffi::{
+    dev_t, gid_t, ino_t, mode_t, rusage, statx, statx_timestamp, timespec, timeval, uid_t, CloneOp,
+    CloseOp, ExitOp, GetRUsageOp, InitProcessOp, InitThreadOp, WaitOp,
+};
+use color_eyre::eyre::{eyre, Context};
+pub use std::ffi::{c_int, c_uint};
+
+use color_eyre::eyre::Result;
+use serde::{Deserialize, Serialize};
+use std::{
+    ffi::{OsStr, OsString},
+    os::unix::ffi::OsStrExt,
+    slice,
+};
+
+use crate::{arena::ArenaContext, ffi};
+
+/// Specialized version of [`std::convert::From`] for working with libprobe arena structs.
+///
+/// Since [`ffi`] structs from arena allocator files have intrinsically invalid pointers (because
+/// they came from a different virtual memory space). This trait and It's sibling [`FfiInto`]
+/// exist to act as [`From`] and [`Into`] with an added parameter of a [`ArenaContext`] that can be
+/// used to decode pointers.
+pub(crate) trait FfiFrom<T> {
+    fn ffi_from(value: &T, ctx: &ArenaContext) -> Result<Self>
+    where
+        Self: Sized;
+}
+
+/// Specialized version of [`std::convert::Into`] for working with libprobe arena structs.
+///
+/// Much like [`std::convert::Into`] this trait is implemented automatically with a blanket
+/// implementation as the reciprocal of [`FfiFrom`].
+pub(crate) trait FfiInto<T> {
+    fn ffi_into(&self, ctx: &ArenaContext) -> Result<T>;
+}
+
+impl<T, U> FfiInto<U> for T
+where
+    U: FfiFrom<T>,
+{
+    #[inline]
+    fn ffi_into(&self, ctx: &ArenaContext) -> Result<U> {
+        U::ffi_from(self, ctx)
+    }
+}
+
+fn try_to_osstring(str: *const i8, ctx: &ArenaContext) -> Result<OsString> {
```

Comment:
```suggestion
/// Most of the strings stored by libprobe are paths
/// POSIX allows paths to have any byte except null, so we can't assume UTF-8 (the normal String)
/// OsString better for this
/// Rust docs say: On Unix systems, strings are often arbitrary sequences of non-zero bytes, in many cases interpreted as UTF-8.
/// https://doc.rust-lang.org/std/ffi/struct.OsString.html
fn try_to_osstring(str: *const i8, ctx: &ArenaContext) -> Result<OsString> {
```

Also, can you comment on using Rust's `OsString` vs Rust's `CString`, considering `CString`s are defined as null-terminated? I think the data coming from libprobe will also be null-terminated. Does it actually matter?

Reply:
To be 100% honest, I forgot that rust had separate `OsString` and `CString` types.

> Does it actually matter? 

On unix platforms, almost definitely not (I believe on *Windows* `OsString`s are UTF-16), but I think the ergonomics of CStrings are slightly better, so I'll refactor accordingly.

Reply:
https://github.com/charmoniumQ/PROBE/pull/20/commits/4b59bd3a473e29481e71d0c24cede2384167ca87